opencog/cogserver/server/AgentRunnerThread.cc

```diff
@@ -0,0 +1,167 @@
+
+using namespace std;
+
+
+namespace opencog
+{
+
+AgentRunnerThread::AgentRunnerThread(const std::string &name) :
+        AgentRunnerBase(name), runAgents(false), agentsModified(false),
+        clearAll(false)
+{
+}
+
+AgentRunnerThread::~AgentRunnerThread()
+{
+    removeAllAgents();
+    joinRunThread();
+}
+
+void AgentRunnerThread::start()
+{
+    {
+        // locking is required, to prevent dead-lock in condition_variable wait()
+        lock_guard<mutex> lock(runAgentsMutex);
+        runAgents.store(true, memory_order_relaxed);
+    }
+    runAgentsCond.notify_one();
+}
+
+void AgentRunnerThread::stop()
+{
+    runAgents.store(false, memory_order_relaxed);
+}
+
+void AgentRunnerThread::addAgent(AgentPtr a)
+{
+    if (!hasAgents()) {
+        joinRunThread();
+        /*
+         * processAgentsThread() will exit immediately if no agents are available,
+         * so we add the first one immediately
+         */
+        {
+            lock_guard<mutex> lock(agentsMutex);
+            AgentRunnerBase::addAgent(a);
+        }
+        runThread = thread(&AgentRunnerThread::processAgentsThread, this);
+        // todo: set thread name when using pthreads
+    }
+    else {
+        lock_guard<mutex> lock(agentsMutex);
+        agentsAddQ.push_back(a);
+        agentsModified.store(true, memory_order_relaxed);
+    }
+}
+
+void AgentRunnerThread::removeAgent(AgentPtr a)
+{
+    lock_guard<mutex> lock(agentsMutex);
+    agentsRemoveQ.push_back(a);
+    agentsModified.store(true, memory_order_relaxed);
+}
+
+void AgentRunnerThread::removeAllAgents(const std::string& id)
+{
+    lock_guard<mutex> lock(agentsMutex);
+    idsRemoveQ.push_back(id);
+    agentsModified.store(true, memory_order_relaxed);
+}
+
+void AgentRunnerThread::removeAllAgents()
+{
+    lock_guard<mutex> lock(agentsMutex);
+    clearAll = true;
+    agentsModified.store(true, memory_order_relaxed);
+}
+
+const AgentSeq& AgentRunnerThread::getAgents() const
+{
+    lock_guard<mutex> lock(agentsMutex);
+    return agents;
+}
+
+bool AgentRunnerThread::hasAgents() const
+{
+    lock_guard<mutex> lock(agentsMutex);
+    return !agents.empty();
+}
+
+void AgentRunnerThread::processAgentsThread()
+{
+    /*
+     * 'agents' is only modified in this function, so there is no need to
+     * protect read accesses to this container in it.
+     */
+    logger().debug("[CogServer::%s] Agent thread started", name.c_str());
+    while (!agents.empty()) {
+        if (!runAgents.load(memory_order_relaxed)) {
+            unique_lock<mutex> lock(runAgentsMutex);
+            runAgentsCond.wait(lock,
+                [this] {return runAgents.load(memory_order_relaxed);});
+        }
+
+
+        AgentSeq::const_iterator it;
+        for (it = agents.begin(); it != agents.end(); ++it) {
+            AgentPtr agent = *it;
+            // todo: can frequency() be still useful?
+//            if ((cycleCount % agent->frequency()) == 0)
+            runAgent(agent);
```

Comment:
OK, this is weird, and my head explodes at this point.  

When an agent is added, using `AgentRunnerThread::addAgent()` you create a new thread, and call this function, `processAgentsThread` as it's body.   I am assuming that the correspondence between threads and agents is one-to-one -- for each thread, there is exactly one agent.  Right?  ??? How could it not be?  so why in the world would you loop over *all* agents, here?  Shouldn't you be working only with that one agent that is yours, and ignore all the other agents running in the other threads? They're none of your business ... why are you messing with them?

And its even more confusing, in yet other ways ... Up above, you have a big while-loop, `while (!agents.empty()) ` and here, we are still inside of that while loop.  Now, given that the threads have all started, and given that they are all stuck waiting in the condition variable, then what, exactly is the intent here?    I mean, fist, you have to unstick the condition variable, so that each thread can move beyond the cond-wait up above.  But, once unstuck, each thread will run.

Ohhh ... I am starting to see... Now, the cond-wait means that only ONE thread can run at a time.  Since only the one holding the conditiion variable can possibly run!  ALL the other threads are waiting, helpless!  Yet here, you call the run method on each agent.  

Now, *if* the agent's run() method is written in such a way that it returns, THEN this loop will advance. Otherwise, this loop will deadlock, because typically, the run() method will never return until that agent is done, and that might be never.  So the only way to not deadlock here is to have each timeslice each agent. 

But this completely defeats the whole purpose of having threads to begin with. You create a bunch of threads -- one for each agent. Then you use a condition variable to make sure that no more than one thread can run at a time. Then when that thread runs, it tries to multi-task the agents. 

Why would you want to create a bunch of threads, and then force them to run single-threaded?   Now, maybe I misunderstand something, but I looked at this code over and over and, from what I can tell, it is doing exactly what I say, up above.  This design makes no sense at all. Its deeply, fundamentally flawed...